src/test/java/UploadReceiptServletTest.java

```diff
@@ -0,0 +1,335 @@
+  private static final String RECEIPT_ANALYSIS_FAILED_WARNING =
+      "com.google.sps.servlets.ReceiptAnalysis$ReceiptAnalysisException: Receipt analysis failed.";
+
+  private static final String INSTANT = "2020-06-22T10:15:30Z";
+
+  private static final long MAX_UPLOAD_SIZE_BYTES = 5 * 1024 * 1024;
+  private static final String UPLOAD_URL = "/blobstore/upload-receipt";
+
+  private static final BlobKey BLOB_KEY = new BlobKey("blobKey");
+  private static final String VALID_FILENAME = "image.jpg";
+  private static final String INVALID_FILENAME = "image.png";
+  private static final long IMAGE_SIZE_1MB = 1024 * 1024;
+  private static final long IMAGE_SIZE_0MB = 0;
+  private static final String HASH = "35454B055CC325EA1AF2126E27707052";
+
+  private static final String LABEL = "Label";
+  private static final Text RAW_TEXT = new Text("raw text");
+  private static final double PRICE = 5.89;
+  private static final String STORE = "McDonald's";
+  private static final AnalysisResults ANALYSIS_RESULTS = new AnalysisResults(RAW_TEXT.getValue());
+
+  private static final String IMAGE_URL = "/serve-image?blob-key=" + BLOB_KEY.getKeyString();
+  private static final String LIVE_SERVER_BASE_URL =
+      "https://capstone-receipt-step-2020.uc.r.appspot.com:80";
+  private static final String LIVE_SERVER_ABSOLUTE_URL = LIVE_SERVER_BASE_URL + IMAGE_URL;
+  private static final String LIVE_SERVER_SCHEME = "https";
+  private static final String LIVE_SERVER_NAME = "capstone-receipt-step-2020.uc.r.appspot.com";
+  private static final int LIVE_SERVER_PORT = 80;
+  private static final String LIVE_SERVER_CONTEXT_PATH = "";
+  private static final String DEV_SERVER_SCHEME = "http";
+  private static final String DEV_SERVER_NAME = "0.0.0.0";
+  private static final int DEV_SERVER_PORT = 80;
+  private static final String DEV_SERVER_CONTEXT_PATH = "";
+
+  @Before
+  public void setUp() {
+    MockitoAnnotations.initMocks(this);
+    helper.setUp();
+    datastore = DatastoreServiceFactory.getDatastoreService();
+
+    // Create a fixed time clock that always returns the same instant.
+    clock = Clock.fixed(Instant.parse(INSTANT), ZoneId.systemDefault());
+
+    servlet = new UploadReceiptServlet(logger, blobstoreService, blobInfoFactory, datastore, clock);
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void doGetReturnsBlobstoreUploadUrl() throws IOException {
+    StringWriter stringWriter = new StringWriter();
+    PrintWriter writer = new PrintWriter(stringWriter);
+    when(response.getWriter()).thenReturn(writer);
+
+    UploadOptions uploadOptions =
+        UploadOptions.Builder.withMaxUploadSizeBytesPerBlob(MAX_UPLOAD_SIZE_BYTES);
+    when(blobstoreService.createUploadUrl("/upload-receipt", uploadOptions)).thenReturn(UPLOAD_URL);
+
+    servlet.doGet(request, response);
+    writer.flush();
+    Assert.assertTrue(stringWriter.toString().contains(UPLOAD_URL));
```

Comment:
We try to structure the test into 3 distinct blocks with a newline between them: arrange, act, assert: every test must arrange the state of the world to test, act on the class under test by calling methods, and assert that the world is in the expected state afterward. So, I would try to group the statements into these 3 blocks.